src/cache_cf.cc

```diff
@@ -1014,87 +1017,133 @@ parse_obsolete(const char *name)
+        debugs(3, DBG_CRITICAL, "FATAL: The time must be less than " << maxYears << " years");
         self_destruct();
     }
 }
 
-static uint64_t
-parseTimeUnits(const char *unit, bool allowMsec)
+/// Parses a time specification from the config file and
+/// returns the time as a chrono duration object of 'TimeUnit' type.
+/// \param defaultUnitName will be used if the time unit is missing,
+/// it should passed only if the configuration paramter being parsed
+/// excpects more arguments after time unit.
+template <class TimeUnit>
+TimeUnit
+parseTimeLine(const char *defaultUnitName = nullptr)
 {
-    if (allowMsec && !strncasecmp(unit, T_MILLISECOND_STR, strlen(T_MILLISECOND_STR)))
-        return 1;
-
-    if (!strncasecmp(unit, T_SECOND_STR, strlen(T_SECOND_STR)))
-        return 1000;
+    char *token = ConfigParser::NextToken();;
+    if (!token) {
+        self_destruct();
+        return TimeUnit::zero();
+    }
 
-    if (!strncasecmp(unit, T_MINUTE_STR, strlen(T_MINUTE_STR)))
-        return 60 * 1000;
+    const auto parsedValue = xatof(token);
 
-    if (!strncasecmp(unit, T_HOUR_STR, strlen(T_HOUR_STR)))
-        return 3600 * 1000;
+    std::chrono::nanoseconds parsedUnitDuration;
 
-    if (!strncasecmp(unit, T_DAY_STR, strlen(T_DAY_STR)))
-        return 86400 * 1000;
+    token = parsedValue ? ConfigParser::PeekAtToken() : nullptr;
 
-    if (!strncasecmp(unit, T_WEEK_STR, strlen(T_WEEK_STR)))
-        return 86400 * 7 * 1000;
+    if (parsedValue && parseTimeUnits<TimeUnit>(token, defaultUnitName, parsedUnitDuration))
+        (void)ConfigParser::NextToken();
 
-    if (!strncasecmp(unit, T_FORTNIGHT_STR, strlen(T_FORTNIGHT_STR)))
-        return 86400 * 14 * 1000;
+    CheckTimeOverflow(parsedValue, parsedUnitDuration);
 
-    if (!strncasecmp(unit, T_MONTH_STR, strlen(T_MONTH_STR)))
-        return static_cast<uint64_t>(86400) * 30 * 1000;
+    const std::chrono::nanoseconds resultDuration(static_cast<std::chrono::nanoseconds::rep>(parsedUnitDuration.count() * parsedValue));
 
-    if (!strncasecmp(unit, T_YEAR_STR, strlen(T_YEAR_STR)))
-        return static_cast<uint64_t>(86400 * 1000 * 365.2522);
+    // validate precisions (time-unit-small only)
+    if (TimeUnit(1) <= std::chrono::microseconds(1) && resultDuration.count() && resultDuration.count() <= 3) {
+        const auto unitName = token ? token : defaultUnitName;
+        assert(unitName);
+        debugs(3, DBG_CRITICAL, "WARNING: The parsed " << parsedValue << " " << unitName <<
+                " on '" << config_input_line << "' is too small to be measured with high_resolution_clock");
+    }
 
-    if (!strncasecmp(unit, T_DECADE_STR, strlen(T_DECADE_STR)))
-        return static_cast<uint64_t>(86400 * 1000 * 365.2522 * 10);
+    const auto result = std::chrono::duration_cast<TimeUnit>(resultDuration);
 
-    debugs(3, DBG_IMPORTANT, "parseTimeUnits: unknown time unit '" << unit << "'");
+    if (parsedValue > 0 && !result.count()) {
```

Comment:
It is not clear why we are testing for positive values here but not above. Perhaps this should be
```suggestion
    if (parsedValue && !result.count()) {
```
but please check that comparing result.count() with zero is how we are supposed to detect underflows.